File: src/main/java/io/github/petrknap/persistence/zoneddatetime/UtcWithLocal.java
package io.github.petrknap.persistence.zoneddatetime;

import jakarta.persistence.Column;
import jakarta.persistence.Embeddable;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.time.LocalDateTime;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;

/**
 * Stores zoned date-time as `utc` date-time with `local` date-time
 */
@Embeddable
public final class UtcWithLocal extends Utc<UtcWithLocal>
{
    @Column(nullable = true)
    private @Nullable LocalDateTime local;

    private UtcWithLocal(@NotNull ZonedDateTime zonedDateTime)
    {
        super(zonedDateTime);
        local = zonedDateTime.toLocalDateTime();
    }

    private UtcWithLocal()
    {
        super();
    }

    public static @NotNull UtcWithLocal of(@NotNull ZonedDateTime zonedDateTime)
    {
        return new UtcWithLocal(zonedDateTime);
    }

    public static @Nullable UtcWithLocal fromStored(
            @Nullable CharSequence utcDateTime,
            @Nullable CharSequence localDateTime,
            @NotNull String dateTimeFormat
    ) {
        return fromStored(
                utcDateTime != null ? DateTimeUtils.parseAsLocalDateTime(utcDateTime, dateTimeFormat) : null,
                localDateTime != null ? DateTimeUtils.parseAsLocalDateTime(localDateTime, dateTimeFormat) : null
        );
    }

    public static @Nullable UtcWithLocal fromStored(
            @Nullable LocalDateTime utcDateTime,
            @Nullable LocalDateTime localDateTime
    ) {
        ZonedDateTime zonedDateTime = ZonedDateTimePersistence.computeZonedDateTime(
                utcDateTime,
                localDateTime
        );

        return zonedDateTime != null ? new UtcWithLocal(zonedDateTime) : null;
    }

    public @NotNull LocalDateTime getLocalDateTime()
    {
        if (local == null) {
            thisInstanceShouldBeNull();
        }
        return local;
    }

    public @NotNull String getLocalDateTime(@NotNull String format)
    {
        return DateTimeFormatter.ofPattern(format).format(getLocalDateTime());
    }

    public @NotNull ZonedDateTime toZonedDateTime()
    {
        return ZonedDateTimePersistence.computeZonedDateTime(
                getUtcDateTime(),
                getLocalDateTime()
        );
    }
}

File: src/main/php/UtcWithLocal.php
<?php

declare(strict_types=1);

namespace PetrKnap\Persistence\ZonedDateTime;

use DateTimeImmutable;
use DateTimeInterface;
use Doctrine\ORM\Mapping as ORM;

/**
 * Stores zoned date-time as `utc` date-time with `local` date-time
 *
 * @phpstan-import-type LocalDateTime from JavaSe8\Time
 */
#[ORM\Embeddable]
final class UtcWithLocal extends Utc
{
    /**
     * @var LocalDateTime|null
     */
    #[ORM\Column(type: 'datetime_immutable', nullable: true)]
    protected DateTimeImmutable|null $local;

    public function __construct(DateTimeInterface $zonedDateTime)
    {
        parent::__construct($zonedDateTime);
        $this->local = JavaSe8\Time::toLocalDateTime(JavaSe8\Time::zonedDateTime($zonedDateTime));
    }

    public static function fromFormattedValues(
        string|null $utcDateTime,
        string|null $localDateTime,
        string $dateTimeFormat,
    ): UtcWithLocal|null {
        return self::fromValues(
            $utcDateTime !== null ? DateTimeUtils::parseAsLocalDateTime($utcDateTime, $dateTimeFormat) : null,
            $localDateTime !== null ? DateTimeUtils::parseAsLocalDateTime($localDateTime, $dateTimeFormat) : null,
        );
    }

    public static function fromValues(
        DateTimeInterface|null $utcDateTime,
        DateTimeInterface|null $localDateTime,
    ): UtcWithLocal|null {
        $zonedDateTime = ZonedDateTimePersistence::computeZonedDateTime(
            $utcDateTime,
            localDateTime: $localDateTime,
        );

        return $zonedDateTime !== null ? new UtcWithLocal($zonedDateTime) : null;
    }

    /**
     * @return ($format is false ? LocalDateTime : string)
     */
    public function getLocalDateTime(string|false $format = false): DateTimeImmutable|string
    {
        if ($this->local === null) {
            $this->thisInstanceShouldBeNull();
        }
        return $format ? $this->local->format($format) : $this->local;
    }

    public function toZonedDateTime(): DateTimeImmutable
    {
        return ZonedDateTimePersistence::computeZonedDateTime(
            $this->getUtcDateTime(),
            localDateTime: $this->getLocalDateTime(),
        );
    }
}
